static-build/posts/2021/07/export-default-thing-vs-thing-as-default/index.md

```diff
@@ -0,0 +1,327 @@
+---
+title: '`export default thing` is different to `export { thing as default }`'
+date: 2021-07-05 01:00:00
+summary: Here's how they differ, and when it matters…
+meta: Here's how they differ, and when it matters…
+image: 'asset-url:./img.png'
+---
+
+[Dominic Elm](https://twitter.com/elmd_) DM'd me on Twitter to ask me questions about circular dependencies, and, well, I didn't know the answer. After some testing, discussion, and _\*ahem\*_ chatting to the V8 team, we figured it out, but I learned something new about JavaScript along the way.
+
+I'm going to leave the circular dependency stuff to the end of the article, as it isn't totally related. First up:
+
+# Imports are references, not values
+
+Here's an import:
+
+```js
+import { thing } from './module.js';
+```
+
+In the above example, `thing` is the same as `thing` in `./module.js`. I know that maybe sounds obvious, but what about:
+
+```js
+const module = await import('./module.js');
+const { thing: destructuredThing } = await import('./module.js');
+```
+
+In this case `module.thing` is the same as `thing` in `./module.js`, whereas `destructuredThing` is a new identifier that's assigned the value of `thing` in `./module.js`, and that behaves differently.
+
+Let's say this is `./module.js`:
+
+```js
+// module.js
+export let thing = 'initial';
+
+setTimeout(() => {
+  thing = 'changed';
+}, 500);
+```
+
+And this is `./main.js`:
+
+```js
+// main.js
+import { thing as importedThing } from './module.js';
+const module = await import('./module.js');
+let { thing } = await import('./module.js');
+
+setTimeout(() => {
+  console.log(importedThing); // "changed"
+  console.log(module.thing); // "changed"
+  console.log(thing); // "initial"
+}, 1000);
+```
+
+Imports are live references, so they pick up the changes. The destructured import doesn't pick up the change because destructuing assigns the current value to a new identifier rather than a live reference.
```

Comment:
```suggestion
Imports are live references, so they pick up the changes. The destructured import doesn't pick up the change because destructuring assigns the current value to a new identifier rather than a live reference.
```